activerecord/test/cases/adapters/postgresql/postgresql_adapter_test.rb

```diff
@@ -484,6 +486,97 @@ def test_only_check_for_insensitive_comparison_capability_once
+        assert_called_with(ActiveRecord::Base.logger, :warn, [mysql_warning]) do
+          @connection.execute("do $$ BEGIN RAISE WARNING 'PostgreSQL SQL warning'; END; $$")
+        end
+      ensure
+        ActiveRecord.db_warnings_action = old_action
+      end
+
+      def test_raises_warnings_when_behaviour_raise
+        old_action = ActiveRecord.db_warnings_action
+        ActiveRecord.db_warnings_action = :raise
+        assert_raises(ActiveRecord::StatementInvalid) do
+          @connection.execute("do $$ BEGIN RAISE WARNING 'PostgreSQL SQL warning'; END; $$")
+        end
+      ensure
+        ActiveRecord.db_warnings_action = old_action
+      end
+
+      def test_reports_when_behaviour_report
+        old_action = ActiveRecord.db_warnings_action
+        ActiveRecord.db_warnings_action = :report
+
+        error_reporter = ActiveSupport::ErrorReporter.new
+        subscriber = ActiveSupport::ErrorReporter::TestHelper::ErrorSubscriber.new
+
+        Rails.define_singleton_method(:error) { error_reporter }
+        Rails.error.subscribe(subscriber)
+
+        @connection.execute("do $$ BEGIN RAISE WARNING 'PostgreSQL SQL warning'; END; $$")
+
+        warning_event, * = subscriber.events.first
+        assert_kind_of ActiveRecord::SqlWarning, warning_event
+        assert_equal "PostgreSQL SQL warning", warning_event.message
+      ensure
+        Rails.singleton_class.remove_method(:error)
+        ActiveRecord.db_warnings_action = old_action
+      end
+
+      def test_warnings_behaviour_can_be_customized_with_a_proc
+        old_action = ActiveRecord.db_warnings_action
+
+        warning_message = nil
+        warning_level = nil
+        ActiveRecord.db_warnings_action = ->(warning) do
+          warning_message = warning.message
+          warning_level = warning.level
+        end
+
+        @connection.execute("do $$ BEGIN RAISE WARNING 'PostgreSQL SQL warning'; END; $$")
+        assert_equal "PostgreSQL SQL warning", warning_message
+        assert_equal "WARNING", warning_level
+      ensure
+        ActiveRecord.db_warnings_action = old_action
+      end
+
+      def test_allowlist_of_warnings_to_ignore
+        old_action = ActiveRecord.db_warnings_action
+        old_ignored_warnings = ActiveRecord.db_warnings_ignore
+        ActiveRecord.db_warnings_action = :raise
+        ActiveRecord.db_warnings_ignore = [/PostgreSQL SQL warning/]
+
+        assert_nothing_raised do
+          @connection.execute("do $$ BEGIN RAISE WARNING 'PostgreSQL SQL warning'; END; $$")
+        end
+      ensure
+        ActiveRecord.db_warnings_action = old_action
+        ActiveRecord.db_warnings_ignore = old_ignored_warnings
+      end
+
+      def test_does_not_raise_notice_level_warnings
+        old_action = ActiveRecord.db_warnings_action
+        ActiveRecord.db_warnings_action = :raise
+        assert_nothing_raised do
+          @connection.execute("DROP TABLE IF EXISTS non_existent_table")
```

Comment:
Can we also assert the return value of `execute` here? I don't love only asserting nothing raised because it could return an unexpected value or the query might not be running at all and we wouldn't know because our only failure scenario is no exception.